### DISCUSSION TITLE
Spiffs filesystem.

### DISCUSSION THREAD

**Discussion by wuftymerguftyguff** [PlummersSoftwareLLC/NightDriverStrip#159] [Q&A]
I am struggling to make the spiffs filesystem at all.

Contents of /data bigger than spiffs partition.

Anyone else having this issue?

**Comment by burginh** [comment]
I'm getting this when I trying to build the file system image, 

Building FS image from 'data' directory to .pio/build/spectrum/spiffs.bin
/index.html
skipping .DS_Store
/asset-manifest.json
skipping .DS_Store
/static/css/main.0f983434.css
/static/js/main.js.gz
SPIFFS_write error(-10001): File system is full.

error adding file!
Error for adding content from js!

is that what your getting?

**Comment by davepl** [comment]
Which config/env is this?  Looks like the flash is full.

**Reply by wuftymerguftyguff** [reply]
For me it is ledstrip, but I added in the web server.  

I think the custom partitions are not right.

I did have a quick look  yesterday, but I need to find some more time.

It is the spiffs portion that is too small, if I remove the gzipped js it builds and flashes the partition

The partition format seems to be setting both size and offset, offset is optional, and will be calculated with correct alignment if offset not specified.

Also sizes can be specified in human readable units. Which does not really matter, but suits more eyes.

**Comment by burginh** [comment]
Spectrum, using the M5StickC-Plus

The data directory looks just under 1MB (mostly the compress java), I have not added anything to the data directory.

**Comment by davepl** [comment]
FWIW, here’s what I’ve used for custom partitions when adding the web server content.  Worked last time I tried it, but I haven’t done so in a while… if this is different than what’s in the tree, give it a shot!

# ESP-IDF Partition Table
# This gives us some additional space for code. 
# It should also fix the OTA regression.  
# Name,   Type, SubType,     Offset,      Size, Flags

nvs,        data,   nvs,     0x009000,    0x005000,
otadata,    data,   ota,     0x00e000,    0x002000,
app0,       app,    ota_0,   0x010000,    0x160000,
app1,       app,    ota_1,   0x170000,    0x160000,
storage,    data,   spiffs,  0x2D0000,    0x130000

**Comment by burginh** [comment]
your numbers match the partition_custom.csv, I cloned the project in the last couple of days so it's pretty fresh.

it is failing during the creation of the spiffs.bin file.

I juggled some files for ballast and figured out the build file systemimage will support 916,260 bytes, if it one byte more it fails.
the data directory from the repo is currently 966,768 bytes (From get info on the data folder).

during the build it outputs this
Wrote 1269520 bytes (755600 compressed) at 0x00010000 in 12.2 seconds (effective 832.5 kbit/s)...

not sure where the extra bytes are coming from

during the program upload there is this line
Flash: [========= ]  87.7% (used 1263753 bytes from 1441792 bytes)
which looks like the flash has been partitioned

**Comment by davepl** [comment]
Are the files in the data directory all compressed with gzip?  I know I had to compress them to get them to fit (and the web server code will unpack them as it serves them).  But they might be checked in uncompressed?

**Comment by burginh** [comment]
there are 5 files in the data folder structure.  Only one appears to be a compressed file.  the rest are tiny in comparison.

data/asset-manifest.json.           222 bytes
data/index.html                            342 bytes
data/css/main.0f983434.css      423 bytes
data/js/main.js.gz                      945KB
data/js/main.js.LIZCENSE.txt.     3KB

Is it an option to trade off the space for the OTA to make the data partition bigger?  That might work until I get to an 8MB ESP32.  I have a ESP32 Feather V2 with 8MB Flash and 2MB PSRAM on its way.  I just want to use the M5Stick when I travel during the holidays with a couple of LED strips.  So I could forgo the OTA if that would help.

**Comment by davepl** [comment]
Yes, if it’s close, I’ve been shaving bytes off the other partitions where possible.  Feel free to do so if you can get a combo that works!

0 Dave

**Comment by burginh** [comment]
I made my own custom partition file, removed the OTA1 and increase the size of the spiffs to 150000.  That allows the build filesystem image to complete.  And I added some code to do a list file systems in the loop every 10 seconds and I see the files listed now.

(I) (loop)(C1) Mem: 104092 LED FPS: 45, LED Watt: 0, LED Brite: 100%, Audio FPS: 38, Serial FPS: 45, PeakVU: 7.34, MinVU: 6.33, VURatio: 0.02, CPU: 52%, 01%, FreeDraw: 0.018182
List All Files
FILE: main.js.gz
FILE: main.js.LICENSE.txt
FILE: index.html
FILE: asset-manifest.json
FILE: main.0f983434.css
(I) (loop)(C1) Mem: 104092 LED FPS: 44, LED Watt: 0, LED Brite: 100%, Audio FPS: 38, Serial FPS: 45, PeakVU: 14.10, MinVU: 7.09, VURatio: 0.01, CPU: 54%, 01%, FreeDraw: 0.018059

When I point chrome at the IP listed it get a web page now.   but all the text is latin looks like lorem ipsum (google translate says latin), there are menus for Home, Effects, Statics and Settings, but they really don't work, some load other text pages of Latin.

So progress...

**Comment by burginh** [comment]
I did some more juggling and got it to work with the two OTA and the SPIFFS storage.  I also learned the the OTA partitions have to be aligned on a 10000 boundary so I couldn't adjust them in smaller amounts.

I ended up with this:
# Name,   Type, SubType,     Offset,      Size, Flags
nvs,        data,   nvs,     0x009000,    0x005000,
otadata,    data,   ota,     0x00e000,    0x002000,
app0,       app,    ota_0,   0x010000,    0x150000,
app1,       app,    ota_1,   0x160000,    0x150000,
storage,    data,   spiffs,  0x2B0000,    0x150000  

The code still fits, but it's tight.
Flash: [========= ]  92.5% (used 1273661 bytes from 1376256 bytes)
and the files fit into the spiff storage.

**Comment by davepl** [comment]
Is this a DEBUG or RELEASE build?   On a 4M flash, we might be limited to a release build if you’re including all of the SPIFFs data, not sure!

**Comment by burginh** [comment]
I was building for release, but I went and tried both. They both fit.

WIth build_type = release
RAM:   [==        ]  19.4% (used 63724 bytes from 327680 bytes)
Flash: [========= ]  92.5% (used 1273661 bytes from 1376256 bytes)
...
Wrote 1279440 bytes (761702 compressed) at 0x00010000 in 11.9 seconds (effective 858.0 kbit/s)...


release has 102595 bytes left in flash
with build_type = debug

RAM:   [==        ]  19.5% (used 63748 bytes from 327680 bytes)
Flash: [==========]  97.7% (used 1344525 bytes from 1376256 bytes)

Wrote 1350304 bytes (799814 compressed) at 0x00010000 in 13.1 seconds (effective 825.8 kbit/s)...

debug has 31731 bytes left in flash.

**Comment by larryl79** [comment]
I had a same issue with env: demo/release and this mod from @burginh helped on it, now I able to upload spiffs without error.
(DOIT ESP32 DEV KIT v1 4mb)

**Reply by erikpt** [reply]
@larryl79 were you able to access the web interface after upload?

**Reply by larryl79** [reply]
for some reason web still inaccessible, I had no time for figure it out why.
